valuable/src/structable.rs

```diff
@@ -7,16 +7,103 @@ pub trait Structable: Valuable {
     fn definition(&self) -> StructDef<'_>;
 }
 
+/// Name, fields, and other struct-level information.
+///
+/// Returned by [`Structable::definition()`], `StructDef` provides the caller
+/// with information about the struct's definition.
+///
+/// [`Structable::definition()`]: Structable::definition
 #[non_exhaustive]
 pub enum StructDef<'a> {
+    /// The struct is statically-defined, all fields are known ahead of time.
+    ///
+    /// # Examples
+    ///
+    /// A statically defined struct
+    ///
+    /// ```
+    /// use valuable::{Fields, Valuable, Structable, StructDef};
+    ///
+    /// #[derive(Valuable)]
+    /// struct MyStruct {
+    ///     foo: &'static str,
+    /// }
+    ///
+    /// let my_struct = MyStruct { foo: "Hello" };
+    /// let fields = match my_struct.definition() {
+    ///     StructDef::Static { name, fields, ..} => {
+    ///         assert_eq!("MyStruct", name);
+    ///         fields
+    ///     }
+    ///     _ => unreachable!(),
+    /// };
+    ///
+    /// match fields {
+    ///     Fields::Named(named_fields) => {
+    ///         assert_eq!(1, named_fields.len());
+    ///         assert_eq!("foo", named_fields[0].name());
+    ///     }
+    ///     _ => unreachable!(),
+    /// }
+    /// ```
     #[non_exhaustive]
     Static {
+        /// The struct's name.
         name: &'static str,
+
+        /// The struct's fields.
         fields: Fields<'static>,
     },
 
+    /// The struct is dynamically-defined, not all fields are known ahead of time.
+    ///
```

Comment:
TIOLI: it could maybe be worth mentioning some potential cases in which you would use this, as opposed to `Mappable`?

the main ones i can think of are:
- wanting to implement `Structable` for foreign non-Rust types that may be dynamic, but indicating that they are the foreign language's "class"/"struct" rather than a foreign language's map/dictionary type
- wanting to mix dynamic and static fields (e.g. i have something like 
   ```rust
   struct Example {
        some_field: Foo,
        some_other_field: Bar,
        extra_fields: HashMap<Whatever, Something>,
   }
   ```
   and i want to treat it as one struct with `some_field` and `some_other_field` plus whatever K/V pairs are in `extra_fields`.

are there any other ones i'm not thinking of here?

Reply:
Those two are the only cases I thought of.